src/server/services/transfers/Allocator.cpp

```diff
@@ -0,0 +1,261 @@
+                                    << commit; 
+
+    /*
+    *   Max flow: run maximum flow algorithm on non-starved links to maximize throughput of allocation
+    */ 
+    
+    // Declare graph variables
+    std::map<std::string, int> seToInt;
+    std::map<int, std::string> intToSe;
+    int nodeCount = 0;
+
+    MaximumFlow::MaximumFlowSolver solver;
+    std::map<std::pair<int, int>, int> maximumFlow;
+
+    // Initialize seToInt and intToSe since our maximum flow class utilizes array indexing for optimization
+    for (const auto& i : queues) {
+        if (seToInt.find(i.sourceSe) == seToInt.end()) {
+            seToInt[i.sourceSe] = nodeCount;
+            intToSe[nodeCount] = i.sourceSe;
+            nodeCount++;
+        }
+        if (seToInt.find(i.destSe) == seToInt.end()) {
+            seToInt[i.destSe] = nodeCount;
+            intToSe[nodeCount] = i.destSe;
+            nodeCount++;
+        }
+    }
+
+    // node count is incremented after each node is added so virtual source is nodeCount, dest is nodeCount + 1
+    solver.setSource(nodeCount);
+    solver.setSink(nodeCount + 1);
+    solver.setNodes(nodeCount + 2);
+
+    // Introduce capacities for virtual source to each source
+    for (const auto& i : queues) {
+        int sourceSlots = slotsLeftForSource[i.sourceSe];
+        int destSlots = slotsLeftForDestination[i.destSe];
+        int linkSlots = linkCapacities[Pair(i.sourceSe, i.destSe)];
+
+        // Reduce graph size by excluding links where the source, destination, or link itself has no slots left, 
+        // and starved links which have already been fully allocated
+        if (sourceSlots > 0 && destSlots > 0 && linkSlots > 0) {
+            solver.addEdge(nodeCount, seToInt[i.sourceSe], sourceSlots);
+            solver.addEdge(seToInt[i.destSe], nodeCount + 1, destSlots);
+            solver.addEdge(seToInt[i.sourceSe], seToInt[i.destSe], linkSlots);
+        }
+    }
+    
+    maximumFlow = solver.computeMaximumFlow();
+    
+    for (const auto& entry : maximumFlow) {
+        Pair link = Pair(intToSe[entry.first.first], intToSe[entry.first.second]);
+        int numSlotsToAllocate = entry.second;
+        allocatorMap[link] = numSlotsToAllocate;
+    }
+
+    // for each allocated link, increment deficit by difference in capacity and allocated flow
+    for (const auto& entry : allocatorMap) {
+        Pair link = entry.first;
+        int numSlotsToAllocate = entry.second;
+        linkDeficits[link] += (linkCapacities[link] - numSlotsToAllocate);
+        // log the total number of slots allocated to this activity by max flow
+        FTS3_COMMON_LOGGER_NEWLOG(INFO) << "J&P&C: "
+                                << "Number of slots allocated (by max flow) for link "
+                                << "source=" << link.source << " dest=" << link.destination << ": "
+                                << numSlotsToAllocate
+                                << commit;
+    }
+
+    return allocatorMap;
```

Comment:
allocatorMap should be created when this function is called and populated when the function is ran